cpp/src/reductions/scan/ewm.cu

```diff
@@ -0,0 +1,424 @@
+        T factor = (1.0 - beta) + pow(beta, nullcnt + 1);
+        return {(beta * (pow(beta, nullcnt)) / factor), ((1.0 - beta) * input) / factor};
+      } else {
+        // value is not valid
+        return this->IDENTITY;
+      }
+    }
+  }
+};
+
+/**
+* @brief Return an array whose values y_i are the number of null entries
+* in between the last valid entry of the input and the current index.
+* Example: {1, NULL, 3, 4, NULL, NULL, 7}
+        -> {0, 0     1, 0, 0,    1,    2}
+*/
+rmm::device_uvector<cudf::size_type> null_roll_up(column_view const& input,
+                                                  rmm::cuda_stream_view stream)
+{
+  rmm::device_uvector<cudf::size_type> output(input.size(), stream);
+
+  auto device_view = column_device_view::create(input);
+  auto valid_it    = cudf::detail::make_validity_iterator(*device_view);
+
+  // Invert the null iterator
+  thrust::transform(rmm::exec_policy(stream),
+                    valid_it,
+                    valid_it + input.size(),
+                    output.begin(),
+                    [] __device__(bool valid) -> bool { return 1 - valid; });
+
+  // 0, 1, 0, 1, 1, 0 -> 0, 0, 1, 0, 0, 2
+  thrust::inclusive_scan_by_key(
+    rmm::exec_policy(stream), output.begin(), output.end() - 1, output.begin(), output.begin() + 1);
+
+  return output;
+}
+
+template <typename T>
+rmm::device_uvector<T> compute_ewma_adjust(column_view const& input,
+                                           T const beta,
+                                           rmm::cuda_stream_view stream,
+                                           rmm::mr::device_memory_resource* mr)
+{
+  rmm::device_uvector<T> output(input.size(), stream);
+  rmm::device_uvector<pair_type<T>> pairs(input.size(), stream);
+  rmm::device_uvector<cudf::size_type> nullcnt(input.size(), stream);
+
+  if (input.has_nulls()) {
+    nullcnt = null_roll_up(input, stream);
+
+    auto device_view = column_device_view::create(input);
+    auto valid_it    = cudf::detail::make_validity_iterator(*device_view);
+    auto data =
+      thrust::make_zip_iterator(thrust::make_tuple(valid_it, nullcnt.begin(), input.begin<T>()));
+
+    thrust::transform_inclusive_scan(rmm::exec_policy(stream),
+                                     data,
+                                     data + input.size(),
+                                     pairs.begin(),
+                                     ewma_adjust_numerator_null_functor{beta},
+                                     recurrence_functor<T>{});
+
+  } else {
+    thrust::transform_inclusive_scan(rmm::exec_policy(stream),
+                                     input.begin<T>(),
+                                     input.end<T>(),
+                                     pairs.begin(),
+                                     ewma_adjust_numerator_nonull_functor{beta},
+                                     recurrence_functor<T>{});
+  }
+
+  // copy the second elements to the output for now
```

Comment:
Looks like this just gets overwritten below?

Reply:
I think in the updates since this comment the code that was overwriting this result was removed. At this point I think this is still needed as this stores the result of the numerator's computation into the output memory, freeing up the array of pairs to be reused for the denominator's computation later on.